boot/boot_serial/src/boot_serial.c

```diff
@@ -485,33 +485,23 @@ bs_upload(char *buf, int len)
     }
 
     BOOT_LOG_INF("Writing at 0x%x until 0x%x", curr_off, curr_off + img_chunk_len);
-    if (rem_bytes) {
+    /* Write flash aligned chunk and if successful and there is rem_bytes to
+     * write, write the too. rem_bytes are only written for the last chunk.
+     */
+    rc = flash_area_write(fap, curr_off, img_chunk, img_chunk_len - rem_bytes);
+    if (rc == 0 && rem_bytes) {
         /* the last chunk of the image might be unaligned */
         uint8_t wbs_aligned[BOOT_MAX_ALIGN];
-        size_t w_size = img_chunk_len - rem_bytes;
 
-        if (w_size) {
-            rc = flash_area_write(fap, curr_off, img_chunk, w_size);
-            if (rc) {
-                goto out_invalid_data;
-            }
-            curr_off += w_size;
-            img_chunk_len -= w_size;
-            img_chunk += w_size;
-        }
+        memset(wbs_aligned, flash_area_erased_val(fap), sizeof(wbs_aligned));
+        memcpy(wbs_aligned, img_chunk + img_chunk_len - rem_bytes, rem_bytes);
 
-        if (img_chunk_len) {
-            memcpy(wbs_aligned, img_chunk, rem_bytes);
-            memset(wbs_aligned + rem_bytes, flash_area_erased_val(fap),
-                   sizeof(wbs_aligned) - rem_bytes);
-            rc = flash_area_write(fap, curr_off, wbs_aligned, flash_area_align(fap));
-        }
-    } else {
-        rc = flash_area_write(fap, curr_off, img_chunk, img_chunk_len);
+        rc = flash_area_write(fap, curr_off + img_chunk_len - rem_bytes,
+                              wbs_aligned, flash_area_align(fap));
     }
 
     if (rc == 0) {
-        curr_off += img_chunk_len;
+        curr_off += img_chunk_len + rem_bytes;
```

Comment:
`curr_off += img_chunk_len;`
Above code always writing img_chunk_len bytes.

Reply:
@de-nordic you missed that. This line patch should be dropped.

Reply:
I have changed logic in 481-485. Now the `rem_bytes` is always subtracted from the `img_chunk_len` so after line 482 `img_chunk_len` holds aligned number of bytes. The `rem_bytes` gets reset to 0 for all writes except the final one.
Therefore the addition done here will only add `rem_bytes` for final write, in all other cases only the aligned `img_chunk_len` is added.

Reply:
looks OK now